README.md
# Kong Prometheus Plugin

[![Build Status][badge-travis-image]][badge-travis-url]

This plugin exposes metrics in [Prometheus Exposition format](https://github.com/prometheus/docs/blob/master/content/docs/instrumenting/exposition_formats.md).


### Available metrics
- *Status codes*: HTTP status codes returned by upstream services. 
- *Latencies Histograms*: Latency as measured at Kong:
   - *Request*: Total request latency 
   - *Kong*: Time taken for Kong to route, authenticate and run all plugins for a request
   - *Upstream*: Time taken by the upstream to respond to the request.
- *Bandwidth*: Total Bandwidth (egress/ingress) flowing through Kong.
- *DB reachability*: Can the Kong node reach it's Database or not (Guage 0/1).
- *Connections*: Various NGINX connection metrics like active, reading, writing,
  accepted connections.

### Grafana Dashboard

Metrics collected via this plugin can be graphed using the following dashboard:
https://grafana.com/dashboards/7424

### Using the plugin

#### Enable the plugin
```bash
$ curl http://localhost:8001/plugins -d name=prometheus
```

### Scraping metrics

#### Via Kong's Admin API

Metrics are available on the admin API at `/metrics` endpoint:
```
curl http://localhost:8001/metrics
```

#### Via Kong's proxy

If your proxy nodes also serve the Admin API, then you can create a route
to `/metrics` endpoint and apply a IP restriction plugin.
```
curl -XPOST http://localhost:8001/services -d name=prometheusEndpoint -d url=http://localhost:8001/metrics
curl -XPOST http://localhost:8001/services/prometheusEndpoint/routes -d paths[]=/metrics
curl -XPOST http://localhost:8001/services/prometheusEndpoint/plugins -d name=ip-restriction -d config.whitelist=10.0.0.0/8
```

#### On a custom port

Alternatively, this plugin has the capability to serve the content on a
different port using a custom server block in Kong's NGINX template.

If you're using Kong 0.14.0 or above, then you can inject the server block
using Kong's [injecting NGINX directives](https://docs.konghq.com/0.14.x/configuration/#injecting-nginx-directives) 
feature.

Consider the below file containing an Nginx `server` block:

```
# /path/to/prometheus-server.conf
server {
    server_name kong_prometheus_exporter;
    listen 0.0.0.0:9542; # can be any other port as well

    location / {
        default_type text/plain;
        content_by_lua_block {
             local prometheus = require "kong.plugins.prometheus.exporter"
             prometheus:collect()
        }
    }

    location /nginx_status {
        internal;
        access_log off;
        stub_status;
    }
}
```

Assuming you've the above file available in your file-system on which
Kong is running, add the following line to your `kong.conf` to scrape metrics
from `9542` port.

```
nginx_http_include=/path/to/prometheus-server.conf
```

If you're running Kong version older than 0.14.0, then you can achieve the
same result by using a
[custom NGINX template](https://docs.konghq.com/0.14.x/configuration/#custom-nginx-templates-embedding-kong).

#### Sample /metrics output

```bash
$ curl http://localhost:8001/metrics
root@vagrant-ubuntu-trusty-64:~# curl -D - http://localhost:8001/metrics
HTTP/1.1 200 OK
Server: openresty/1.11.2.5
Date: Mon, 11 Jun 2018 01:39:38 GMT
Content-Type: text/plain; charset=UTF-8
Transfer-Encoding: chunked
Connection: keep-alive
Access-Control-Allow-Origin: *

# HELP kong_bandwidth Total bandwidth in bytes consumed per service in Kong
# TYPE kong_bandwidth counter
kong_bandwidth{type="egress",service="google"} 1277
kong_bandwidth{type="ingress",service="google"} 254
# HELP kong_datastore_reachable Datastore reachable from Kong, 0 is unreachable
# TYPE kong_datastore_reachable gauge
kong_datastore_reachable 1
# HELP kong_http_consumer_status HTTP status codes for customer per service/route in Kong
# TYPE kong_http_consumer_status counter
kong_http_consumer_status{service="upstream",route="default",code="200",consumer="consumer1"} 5185
# HELP kong_http_status HTTP status codes per service in Kong
# TYPE kong_http_status counter
kong_http_status{code="301",service="google"} 2
# HELP kong_http_url_location_consumer_total HTTP status codes for specific URL location in Kong
# TYPE kong_http_url_location_consumer_total counter
kong_http_url_location_consumer_total{service="upstream",route="default",location="hello",consumer="consumer1"} 5
# HELP kong_http_url_param_consumer_total HTTP status codes for specific GET param in Kong
# TYPE kong_http_url_param_consumer_total counter
kong_http_url_param_consumer_total{service="upstream",route="default",param="123456",consumer="consumer1"} 5
# HELP kong_latency Latency added by Kong, total request time and upstream latency for each service in Kong
# TYPE kong_latency histogram
kong_latency_bucket{type="kong",service="google",le="00001.0"} 1
kong_latency_bucket{type="kong",service="google",le="00002.0"} 1
.
.
.
kong_latency_bucket{type="kong",service="google",le="+Inf"} 2
kong_latency_bucket{type="request",service="google",le="00300.0"} 1
kong_latency_bucket{type="request",service="google",le="00400.0"} 1
.
.
kong_latency_bucket{type="request",service="google",le="+Inf"} 2
kong_latency_bucket{type="upstream",service="google",le="00300.0"} 2
kong_latency_bucket{type="upstream",service="google",le="00400.0"} 2
.
.
kong_latency_bucket{type="upstream",service="google",le="+Inf"} 2
kong_latency_count{type="kong",service="google"} 2
kong_latency_count{type="request",service="google"} 2
kong_latency_count{type="upstream",service="google"} 2
kong_latency_sum{type="kong",service="google"} 2145
kong_latency_sum{type="request",service="google"} 2672
kong_latency_sum{type="upstream",service="google"} 527
# HELP kong_nginx_http_current_connections Number of HTTP connections
# TYPE kong_nginx_http_current_connections gauge
kong_nginx_http_current_connections{state="accepted"} 8
kong_nginx_http_current_connections{state="active"} 1
kong_nginx_http_current_connections{state="handled"} 8
kong_nginx_http_current_connections{state="reading"} 0
kong_nginx_http_current_connections{state="total"} 8
kong_nginx_http_current_connections{state="waiting"} 0
kong_nginx_http_current_connections{state="writing"} 1
# HELP kong_nginx_metric_errors_total Number of nginx-lua-prometheus errors
# TYPE kong_nginx_metric_errors_total counter
kong_nginx_metric_errors_total 0
# HELP kong_upstream_target_health Health status of targets of upstream. States = healthchecks_off|healthy|unhealthy|dns_error, value is 1 when state is populated.
kong_upstream_target_health{upstream="<upstream_name>",target="<target>",address="<ip>:<port>",state="healthchecks_off"} 0
kong_upstream_target_health{upstream="<upstream_name>",target="<target>",address="<ip>:<port>",state="healthy"} 1
kong_upstream_target_health{upstream="<upstream_name>",target="<target>",address="<ip>:<port>",state="unhealthy"} 0
kong_upstream_target_health{upstream="<upstream_name>",target="<target>",address="<ip>:<port>",state="dns_error"} 0

```



[badge-travis-url]: https://travis-ci.com/Kong/kong-plugin-prometheus/branches
[badge-travis-image]: https://travis-ci.com/Kong/kong-plugin-prometheus.svg?branch=master


CHANGELOG.md
# Table of Contents

- [1.2.2](#122---20210609)
- [1.2.1](#121---20210415)
- [1.2.0](#120---20210324)
- [1.1.0](#110---20210303)
- [1.0.0](#100---20200820)
- [0.9.0](#090---20200617)
- [0.8.0](#080---20200424)
- [0.7.1](#071---20200105)
- [0.7.0](#070---20191204)
- [0.6.0](#060---20190929)
- [0.5.0](#050---20190916)
- [0.4.1](#041---20190801)
- [0.4.0](#040---20190605)
- [0.3.4](#034---20181217)
- [0.3.3](#033---20181214)
- [0.3.2](#032---20181101)
- [0.3.1](#031---20181017)
- [0.3.0](#030---20181015)
- [0.2.0](#020---20180924)
- [0.1.0](#010---20180615)

##  [1.2.2] - 2021/06/9

- Fix exporter to attach subsystem label to memory stats
  [#118](https://github.com/Kong/kong-plugin-prometheus/pull/118)


##  [1.2.1] - 2021/04/15

- Fix an issue where the Content-Length header could be potentially mispelled
  [#124](https://github.com/Kong/kong-plugin-prometheus/pull/124)

##  [1.2.0] - 2021/03/24

- Fix an issue where there's no stream listener or stream API is not available,
/metrics endpoint may timeout [#108](https://github.com/Kong/kong-plugin-prometheus/pull/108)
- Export per-consumer status [#115](https://github.com/Kong/kong-plugin-prometheus/pull/115)
(Thanks, [samsk](https://github.com/samsk)!)

##  [1.1.0] - 2021/03/03

- Export Kong Enterprise Edition licensing information.
  [#110](https://github.com/Kong/kong-plugin-prometheus/pull/110)

##  [1.0.0] - 2020/08/20

- Change handler to use Kong PDK function kong.log.serialize instead of using
  a deprecated basic serializer.

##  [0.9.0] - 2020/06/17

- Expose healthiness of upstream targets
  (Thanks, [carnei-ro](https://github.com/carnei-ro)!)
  [#88](https://github.com/Kong/kong-plugin-prometheus/pull/88)
- Fix a typo on the dashboard
  (Thanks, [Monska85](https://github.com/Monska85)!)

##  [0.8.0] - 2020/04/24

- Expose the `prometheus` object for custom metrics
  [#78](https://github.com/Kong/kong-plugin-prometheus/pull/78)
- Significant performance enhancements; expect manifolds improvements in
  Kong's throughput while using the plugin and reduction in CPU usage while
  memory usage is expected to go up.
  [#79](https://github.com/Kong/kong-plugin-prometheus/pull/79)

##  [0.7.1] - 2020/01/05

- Fix `full_metric_name` function was not accessible
- Fix linting issues

##  [0.7.0] - 2019/12/04

- **Performance improvements:** Reduced the number of writes (and hence locks)
  to the shared dictionary using lua-resty-counter library.
  (Status API is being shipped with Kong 1.4).
  [#69](https://github.com/Kong/kong-plugin-prometheus/pull/69)
- Update schema for the plugin for Kong 2.0 compatibility
  [#72](https://github.com/Kong/kong-plugin-prometheus/pull/72)

##  [0.6.0] - 2019/09/29

- **Metrics on Status API:** Metrics are now be available on the Status API
  (Status API is being shipped with Kong 1.4).
  [#66](https://github.com/Kong/kong-plugin-prometheus/pull/66)

##  [0.5.0] - 2019/09/16

- **Route based metrics:**  All proxy metrics now contain a tag with the name
  or ID of the route.
  [#40](https://github.com/Kong/kong-plugin-prometheus/issues/40)
- **New metrics releated to Kong's memory usage:**
  New metrics related to Kong's shared dictionaries
  and Lua VMs are now available
  [#62](https://github.com/Kong/kong-plugin-prometheus/pull/62):
  - per worker Lua VM allocated bytes (`kong_memory_workers_lua_vms_bytes`)
  - shm capacity and bytes allocated (`kong_memory_lua_shared_dict_bytes` and
    `kong_memory_lua_shared_dict_total_bytes`)
- Performance has been improved by avoiding unnecessary timer creation.
  This will lower the impact of the plugin on Kong's overall latency.
  [#60](https://github.com/Kong/kong-plugin-prometheus/pull/60)
- Tests to ensure gRPC compatibility have been added.
  [#57](https://github.com/Kong/kong-plugin-prometheus/pull/57)

##  [0.4.1] - 2019/08/01

- Fix issue where the plugin's shared dictionary would not be properly
initialized

##  [0.4.0] - 2019/06/05

- Remove BasePlugin inheritance (not needed anymore)

##  [0.3.4] - 2018/12/17

- Drop the use of `kong.tools.responses` module for
  Kong 1.0 compatibility.
  [#34](https://github.com/Kong/kong-plugin-prometheus/pull/34)

##  [0.3.3] - 2018/12/14

- Do not attempt to send HTTP status code after the body has been sent
  while serving `/metrics`. This would result in error being logged in Kong.
  [#33](https://github.com/Kong/kong-plugin-prometheus/pull/33)

##  [0.3.2] - 2018/11/01

- Fix a nil pointer de-reference bug when no routes are matched in Kong.
  [#28](https://github.com/Kong/kong-plugin-prometheus/pull/28)

##  [0.3.1] - 2018/10/17

- Fix bugs introduced in 0.3.0 due to incorrect PDK function calls
  Thank you @kikito for the fix!
  [#26](https://github.com/Kong/kong-plugin-prometheus/pull/26)

##  [0.3.0] - 2018/10/15

- This release has no user facing changes but has under the hood
  changes for upcoming Kong 1.0.0 release.
- Migrated schema and API endpoint of the plugin to the new DAO and
  use PDK functions where possible.
  Thank you @kikito for the contribution!
  [#24](https://github.com/Kong/kong-plugin-prometheus/pull/24)

##  [0.2.0] - 2018/09/24

- :warning: Dropped metrics that were aggregated across services in Kong.
  These metrics can be obtained much more efficiently using queries in Prometheus.
  [#8](https://github.com/Kong/kong-plugin-prometheus/pull/8)

##  [0.1.0] - 2018/06/15

- Initial release of Prometheus plugin for Kong.

[1.2.2]: https://github.com/Kong/kong-plugin-prometheus/compare/1.2.1...1.2.2
[1.2.1]: https://github.com/Kong/kong-plugin-prometheus/compare/1.2.0...1.2.1
[1.2.0]: https://github.com/Kong/kong-plugin-prometheus/compare/1.1.0...1.2.0
[1.1.0]: https://github.com/Kong/kong-plugin-prometheus/compare/1.0.0...1.1.0
[1.0.0]: https://github.com/Kong/kong-plugin-prometheus/compare/0.9.0...1.0.0
[0.9.0]: https://github.com/Kong/kong-plugin-prometheus/compare/0.8.0...0.9.0
[0.8.0]: https://github.com/Kong/kong-plugin-prometheus/compare/0.7.1...0.8.0

kong/plugins/prometheus/exporter.lua
local kong = kong
local ngx = ngx
local find = string.find
local lower = string.lower
local concat = table.concat
local select = select
local balancer = require("kong.runloop.balancer")

local CLUSTERING_SYNC_STATUS = require("kong.constants").CLUSTERING_SYNC_STATUS

local stream_available, stream_api = pcall(require, "kong.tools.stream_api")

local role = kong.configuration.role

local DEFAULT_BUCKETS = { 1, 2, 5, 7, 10, 15, 20, 25, 30, 40, 50, 60, 70,
                          80, 90, 100, 200, 300, 400, 500, 1000,
                          2000, 5000, 10000, 30000, 60000 }
local metrics = {}
-- prometheus.lua instance
local prometheus

-- the clustering module, lazily imported
local clustering

-- use the same counter library shipped with Kong
package.loaded['prometheus_resty_counter'] = require("resty.counter")

local enterprise
local pok = pcall(require, "kong.enterprise_edition.licensing")
if pok then
  enterprise = require("kong.plugins.prometheus.enterprise.exporter")
end

local kong_subsystem = ngx.config.subsystem

local function init()
  local shm = "prometheus_metrics"
  if not ngx.shared.prometheus_metrics then
    kong.log.err("prometheus: ngx shared dict 'prometheus_metrics' not found")
    return
  end

  prometheus = require("kong.plugins.prometheus.prometheus").init(shm, "kong_")

  -- global metrics
  if kong_subsystem == "http" then
    metrics.connections = prometheus:gauge("nginx_http_current_connections",
      "Number of HTTP connections",
      {"state"})
  else
    metrics.connections = prometheus:gauge("nginx_stream_current_connections",
      "Number of Stream connections",
      {"state"})
  end
  metrics.db_reachable = prometheus:gauge("datastore_reachable",
                                          "Datastore reachable from Kong, " ..
                                          "0 is unreachable")
  metrics.upstream_target_health = prometheus:gauge("upstream_target_health",
                                          "Health status of targets of upstream. " ..
                                          "States = healthchecks_off|healthy|unhealthy|dns_error, " ..
                                          "value is 1 when state is populated.",
                                          {"upstream", "target", "address", "state"})

  local memory_stats = {}
  memory_stats.worker_vms = prometheus:gauge("memory_workers_lua_vms_bytes",
                                             "Allocated bytes in worker Lua VM",
                                             {"pid", "kong_subsystem"})
  memory_stats.shms = prometheus:gauge("memory_lua_shared_dict_bytes",
                                       "Allocated slabs in bytes in a shared_dict",
                                       {"shared_dict", "kong_subsystem"})
  memory_stats.shm_capacity = prometheus:gauge("memory_lua_shared_dict_total_bytes",
                                               "Total capacity in bytes of a shared_dict",
                                               {"shared_dict", "kong_subsystem"})

  local res = kong.node.get_memory_stats()
  for shm_name, value in pairs(res.lua_shared_dicts) do
    memory_stats.shm_capacity:set(value.capacity, { shm_name, kong_subsystem })
  end

  metrics.memory_stats = memory_stats

  -- per service/route
  if kong_subsystem == "http" then
    metrics.status = prometheus:counter("http_status",
                                        "HTTP status codes per service/route in Kong",
                                        {"service", "route", "code"})
  else
    metrics.status = prometheus:counter("stream_status",
                                        "Stream status codes per service/route in Kong",
                                        {"service", "route", "code"})
  end
  metrics.latency = prometheus:histogram("latency",
                                         "Latency added by Kong, total " ..
                                         "request time and upstream latency " ..
                                         "for each service/route in Kong",
                                         {"service", "route", "type"},
                                         DEFAULT_BUCKETS) -- TODO make this configurable
  metrics.bandwidth = prometheus:counter("bandwidth",
                                         "Total bandwidth in bytes " ..
                                         "consumed per service/route in Kong",
                                         {"service", "route", "type"})
  metrics.consumer_status = prometheus:counter("http_consumer_status",
                                          "HTTP status codes for customer per service/route in Kong",
                                          {"service", "route", "code", "consumer"})

  if enterprise then
    enterprise.init(prometheus)
  end


  -- Hybrid mode status
  if role == "control_plane" then
    clustering = require("kong.clustering")

    metrics.data_plane_last_seen = prometheus:gauge("data_plane_last_seen",
                                              "Last time data plane contacted control plane",
                                              {"node_id", "hostname", "ip"})
    metrics.data_plane_config_hash = prometheus:gauge("data_plane_config_hash",
                                              "Config hash numeric value of the data plane",
                                              {"node_id", "hostname", "ip"})

    metrics.data_plane_version_compatible = prometheus:gauge("data_plane_version_compatible",
                                              "Version compatible status of the data plane, 0 is incompatible",
                                              {"node_id", "hostname", "ip", "kong_version"})
  end
end

local function init_worker()
  prometheus:init_worker()
end

-- Convert the MD5 hex string to its numeric representation
-- Note the following will be represented as a float instead of int64 since luajit
-- don't like int64. Good news is prometheus uses float instead of int64 as well
local function config_hash_to_number(hash_str)
  return tonumber("0x" .. hash_str)
end

-- Since in the prometheus library we create a new table for each diverged label
-- so putting the "more dynamic" label at the end will save us some memory
local labels_table = {0, 0, 0}
local labels_table4 = {0, 0, 0, 0}
local upstream_target_addr_health_table = {
  { value = 0, labels = { 0, 0, 0, "healthchecks_off" } },
  { value = 0, labels = { 0, 0, 0, "healthy" } },
  { value = 0, labels = { 0, 0, 0, "unhealthy" } },
  { value = 0, labels = { 0, 0, 0, "dns_error" } },
}

local function set_healthiness_metrics(table, upstream, target, address, status, metrics_bucket)
  for i = 1, #table do
    table[i]['labels'][1] = upstream
    table[i]['labels'][2] = target
    table[i]['labels'][3] = address
    table[i]['value'] = (status == table[i]['labels'][4]) and 1 or 0
    metrics_bucket:set(table[i]['value'], table[i]['labels'])
  end
end


local log

if kong_subsystem == "http" then
  function log(message, serialized)
    if not metrics then
      kong.log.err("prometheus: can not log metrics because of an initialization "
              .. "error, please make sure that you've declared "
              .. "'prometheus_metrics' shared dict in your nginx template")
      return
    end

    local service_name
    if message and message.service then
      service_name = message.service.name or message.service.host
    else
      -- do not record any stats if the service is not present
      return
    end

    local route_name
    if message and message.route then
      route_name = message.route.name or message.route.id
    end

    labels_table[1] = service_name
    labels_table[2] = route_name
    labels_table[3] = message.response.status
    metrics.status:inc(1, labels_table)

    local request_size = tonumber(message.request.size)
    if request_size and request_size > 0 then
      labels_table[3] = "ingress"
      metrics.bandwidth:inc(request_size, labels_table)
    end

    local response_size = tonumber(message.response.size)
    if response_size and response_size > 0 then
      labels_table[3] = "egress"
      metrics.bandwidth:inc(response_size, labels_table)
    end

    local request_latency = message.latencies.request
    if request_latency and request_latency >= 0 then
      labels_table[3] = "request"
      metrics.latency:observe(request_latency, labels_table)
    end

    local upstream_latency = message.latencies.proxy
    if upstream_latency ~= nil and upstream_latency >= 0 then
      labels_table[3] = "upstream"
      metrics.latency:observe(upstream_latency, labels_table)
    end

    local kong_proxy_latency = message.latencies.kong
    if kong_proxy_latency ~= nil and kong_proxy_latency >= 0 then
      labels_table[3] = "kong"
      metrics.latency:observe(kong_proxy_latency, labels_table)
    end

    if serialized.consumer ~= nil then
      labels_table4[1] = labels_table[1]
      labels_table4[2] = labels_table[2]
      labels_table4[3] = message.response.status
      labels_table4[4] = serialized.consumer
      metrics.consumer_status:inc(1, labels_table4)
    end
  end

else
  function log(message)
    if not metrics then
      kong.log.err("prometheus: can not log metrics because of an initialization "
              .. "error, please make sure that you've declared "
              .. "'prometheus_metrics' shared dict in your nginx template")
      return
    end

    local service_name
    if message and message.service then
      service_name = message.service.name or message.service.host
    else
      -- do not record any stats if the service is not present
      return
    end

    local route_name
    if message and message.route then
      route_name = message.route.name or message.route.id
    end

    labels_table[1] = service_name
    labels_table[2] = route_name
    labels_table[3] = message.session.status
    metrics.status:inc(1, labels_table)

    local ingress_size = tonumber(message.session.received)
    if ingress_size and ingress_size > 0 then
      labels_table[3] = "ingress"
      metrics.bandwidth:inc(ingress_size, labels_table)
    end

    local egress_size = tonumber(message.session.sent)
    if egress_size and egress_size > 0 then
      labels_table[3] = "egress"
      metrics.bandwidth:inc(egress_size, labels_table)
    end

    local session_latency = message.latencies.session
    if session_latency and session_latency >= 0 then
      labels_table[3] = "request"
      metrics.latency:observe(session_latency, labels_table)
    end

    local kong_proxy_latency = message.latencies.kong
    if kong_proxy_latency ~= nil and kong_proxy_latency >= 0 then
      labels_table[3] = "kong"
      metrics.latency:observe(kong_proxy_latency, labels_table)
    end
  end
end


local function metric_data()
  if not prometheus or not metrics then
    kong.log.err("prometheus: plugin is not initialized, please make sure ",
                 " 'prometheus_metrics' shared dict is present in nginx template")
    return kong.response.exit(500, { message = "An unexpected error occurred" })
  end

  if ngx.location then
    local r = ngx.location.capture "/nginx_status"

    if r.status ~= 200 then
      kong.log.warn("prometheus: failed to retrieve /nginx_status ",
        "while processing /metrics endpoint")

    else
      local accepted, handled, total = select(3, find(r.body,
        "accepts handled requests\n (%d*) (%d*) (%d*)"))
      metrics.connections:set(accepted, { "accepted" })
      metrics.connections:set(handled, { "handled" })
      metrics.connections:set(total, { "total" })
    end
  end

  metrics.connections:set(ngx.var.connections_active or 0, { "active" })
  metrics.connections:set(ngx.var.connections_reading or 0, { "reading" })
  metrics.connections:set(ngx.var.connections_writing or 0, { "writing" })
  metrics.connections:set(ngx.var.connections_waiting or 0, { "waiting" })

  -- db reachable?
  local ok, err = kong.db.connector:connect()
  if ok then
    metrics.db_reachable:set(1)

  else
    metrics.db_reachable:set(0)
    kong.log.err("prometheus: failed to reach database while processing",
                 "/metrics endpoint: ", err)
  end

  -- erase all target/upstream metrics, prevent exposing old metrics
  metrics.upstream_target_health:reset()

  -- upstream targets accessible?
  local upstreams_dict = balancer.get_all_upstreams()
  for key, upstream_id in pairs(upstreams_dict) do
    local _, upstream_name = key:match("^([^:]*):(.-)$")
    upstream_name = upstream_name and upstream_name or key
    -- based on logic from kong.db.dao.targets
    local health_info
    health_info, err = balancer.get_upstream_health(upstream_id)
    if err then
      kong.log.err("failed getting upstream health: ", err)
    end

    if health_info then
      for target_name, target_info in pairs(health_info) do
        if target_info ~= nil and target_info.addresses ~= nil and
          #target_info.addresses > 0 then
          -- healthchecks_off|healthy|unhealthy
          for _, address in ipairs(target_info.addresses) do
            local address_label = concat({address.ip, ':', address.port})
            local status = lower(address.health)
            set_healthiness_metrics(upstream_target_addr_health_table, upstream_name, target_name, address_label, status, metrics.upstream_target_health)
          end
        else
          -- dns_error
          set_healthiness_metrics(upstream_target_addr_health_table, upstream_name, target_name, '', 'dns_error', metrics.upstream_target_health)
        end
      end
    end
  end

  -- memory stats
  local res = kong.node.get_memory_stats()
  for shm_name, value in pairs(res.lua_shared_dicts) do
    metrics.memory_stats.shms:set(value.allocated_slabs, { shm_name, kong_subsystem })
  end
  for i = 1, #res.workers_lua_vms do
    metrics.memory_stats.worker_vms:set(res.workers_lua_vms[i].http_allocated_gc,
                                        { res.workers_lua_vms[i].pid, kong_subsystem })
  end

  if enterprise then
    enterprise.metric_data()
  end

  -- Hybrid mode status
  if role == "control_plane" then
    -- Cleanup old metrics
    metrics.data_plane_last_seen:reset()
    metrics.data_plane_config_hash:reset()

    for data_plane, err in kong.db.clustering_data_planes:each() do
      if err then
        kong.log.err("failed to list data planes: ", err)
        goto next_data_plane
      end

      local labels = { data_plane.id, data_plane.hostname, data_plane.ip }

      metrics.data_plane_last_seen:set(data_plane.last_seen, labels)
      metrics.data_plane_config_hash:set(config_hash_to_number(data_plane.config_hash), labels)

      labels[4] = data_plane.version
      local compatible = 1

      if data_plane.sync_status == CLUSTERING_SYNC_STATUS.KONG_VERSION_INCOMPATIBLE
        or data_plane.sync_status == CLUSTERING_SYNC_STATUS.PLUGIN_SET_INCOMPATIBLE
        or data_plane.sync_status == CLUSTERING_SYNC_STATUS.PLUGIN_VERSION_INCOMPATIBLE then

        compatible = 0
      end
      metrics.data_plane_version_compatible:set(compatible, labels)

::next_data_plane::
    end
  end

  return prometheus:metric_data()
end

local function collect(with_stream)
  ngx.header["Content-Type"] = "text/plain; charset=UTF-8"

  ngx.print(metric_data())

  -- only gather stream metrics if stream_api module is avaiable
  -- and user has configured at least one stream listeners
  if stream_available and #kong.configuration.stream_listeners > 0 then
    local res, err = stream_api.request("prometheus", "")
    if err then
      kong.log.err("failed to collect stream metrics: ", err)
    else
      ngx.print(res)
    end
  end
end

local function get_prometheus()
  if not prometheus then
    kong.log.err("prometheus: plugin is not initialized, please make sure ",
                     " 'prometheus_metrics' shared dict is present in nginx template")
  end
  return prometheus
end

return {
  init        = init,
  init_worker = init_worker,
  log         = log,
  metric_data = metric_data,
  collect     = collect,
  get_prometheus = get_prometheus,
}
